### PR TITLE
Refactor Adobe AEM hash processing script

### PR DESCRIPTION
**Pull request by tsvenbla**
This commit applies several improvements to the Adobe AEM hash processing script. The changes enhance code readability, improve error handling, and ensure compatibility with both Python 2 and Python 3.

1. Improved readability and modularity: The original monolithic `process_file` function was split into several smaller functions (`split_user_and_hash`, `extract_algo_and_hash`, `split_hash_parts`, and `process_line`). This separation of concerns makes the code easier to understand and maintain.

2. Enhanced error handling: Instead of printing error messages directly to stderr, the updated script uses Python's `logging` module. This provides more control over error reporting and makes adding additional information in log messages easier.

3. Ensured Python 2/3 compatibility: The script now checks the Python version at runtime and adjusts its behavior accordingly. This includes using the `print` function (instead of the `print` statement) and handling file paths correctly across Python versions. And as I already mentioned in [tsvenbla:patch-2](https://github.com/openwall/john/pull/5321), we should probably phase out Python 2 as it has been EOL since 2020.

4. Replaced manual argument parsing with argparse: The updated script uses Python's `argparse` module to handle command-line arguments. This provides a more robust and user-friendly interface, automatically generating help and usage messages.

5. Replaced hardcoded algorithm tags with a dictionary: The tags and their respective algorithm numbers are now stored in a dictionary (`ALGO_TAGS`). This makes it easier to add support for additional hash algorithms in the future.

### CODE REVIEW

**Review by ghost — APPROVED** [review]
It is a 2john tool. We should allow more people to use and test it. We trust your testing. 

I'll try to test it (if we have samples in john-samples). If you have goods samples, please create a PR at https://github.com/openwall/john-samples.

File: run/aem2john.py
Comment (solardiz) [inline run/aem2john.py:24]: This wording leaves the current license undefined. Instead of rewording the previous license statement, can you please keep it intact and add your own copyright statement along with Dhiru's? That way, both of you would be copyright holders and the license would remain as it were.

Also, Claudio managed to confuse you about line lengths. Sorry! For comments, let's keep them under 80 (so 79 max). This means you did not need to re-wrap the previously existing lines (Dhiru's) and you need to make yours consistent with those.

(For non-comment code lines, we can let them occasionally go up to 131 where needed.)

Thank you, and sorry for the delay in reviewing your PRs. I intend to proceed quicker now.

**Review by solardiz — CHANGES_REQUESTED** [review]

**Comment by solardiz** [comment]
@exploide @kholia We'd appreciate your review. Thank you!

**Comment by solardiz** [comment]
@tsvenbla Can you please share some context on your making changes to these scripts in general - not only this one, but your series of PRs? Specifically, are you using those scripts yourself, or are you just trying to help the project by cleaning up code that you don't even use? Also, is the alphabetic order just a coincidence or are you going to refactor the scripts (eventually all of them?) in that order? Regardless of your answers to these, I greatly appreciate your willingness to contribute. I just felt it'd help us all to see the bigger picture and plan accordingly. Thank you!

File: run/aem2john.py
Comment (exploide) [inline run/aem2john.py:92]: This prints the `:` even if `user` is empty, resulting in an invalid hash.

File: run/aem2john.py
Comment (exploide) [inline run/aem2john.py:82]: I love f-strings too, but the script claims to be python2 compatible and f-strings are python3.6+

File: run/aem2john.py
Comment (exploide) [inline run/aem2john.py:24]: The example hashes are gone in your updated version. I think it is really useful to see how expected input data looks like. Maybe keep them somewhere. I guess after the "Each argument should be a file path pointing to a text file containing AEM hashes, one per line." comment above would be a good place.

File: run/aem2john.py
Comment (exploide) [inline run/aem2john.py:109]: Maybe some more precise description like "Convert Adobe AEM hashes to the format suitable for John the Ripper"

**Review by exploide — CHANGES_REQUESTED** [review]
I endorse this PR in general. The code is more comprehensible and more idiomatic Python. However, I left a few review comments.

Concerning Python 2, I feel your desire to abandon this version as it is EOL and makes maintenance unnecessarily hard. I also had this discussion with @solardiz some time ago ;) However, the maintainers decide.

File: run/aem2john.py
Comment (solardiz) [reply run/aem2john.py:92]: I see no problem with printing the `:` when `user` is empty - if the line starts with a `:`, then that indicates the username is empty. I think that's how it should be.

File: run/aem2john.py
Comment (solardiz) [reply run/aem2john.py:82]: Oh, I guess this implies this revised script wasn't tested on Python below 3.6 at all, and the Python 2 compatibility is purely theoretical? I think we shouldn't be merging these PRs without actual testing across a range of Python versions.

**Comment by solardiz** [comment]
Thank you very much for the review, @exploide! I think we should now wait for @tsvenbla's answer about the big picture, then decide whether/how we want to proceed.

File: run/aem2john.py
Comment (exploide) [reply run/aem2john.py:92]: Just noted this because it's a difference between the old and the new behavior. But you are right. The hash with the leading colon is indeed crackable. So this comment could be ignored.

**Comment by solardiz** [comment]
@tsvenbla Hey, why did you end up deleting the repo / closing these PRs? I was expecting you'd comment on my "big picture" comment/question. If we ended up discouraging you, I am really sorry about that!

**Comment by tsvenbla** [comment]
Oooh, so that’s how it works. As you may gather, I am quite new to GitHub. Still learning. How should I proceed from here? I am only doing this on my spare time, so I am not as active as I would like.

**Comment by solardiz** [comment]
@tsvenbla Thanks for your reply. I suggest that you reply to my comment https://github.com/openwall/john/pull/5322#issuecomment-1630741149 and then we see how to proceed.

**Comment by tsvenbla** [comment]
> @tsvenbla Can you please share some context on your making changes to these scripts in general - not only this one, but your series of PRs? Specifically, are you using those scripts yourself, or are you just trying to help the project by cleaning up code that you don't even use? Also, is the alphabetic order just a coincidence or are you going to refactor the scripts (eventually all of them?) in that order? Regardless of your answers to these, I greatly appreciate your willingness to contribute. I just felt it'd help us all to see the bigger picture and plan accordingly. Thank you!

I was aiming to refactor the scripts in order. No specific goal in mind, just contributing where I see improvement!
I also uploaded some tests in the repo john-sample to ensure it works correctly. Not sure what happened to it now that I deleted my fork though. I didn't know it would close the PRs.

**Pull request closed without merging**
